woocommerce/woocommerce
# Enhance remote logging data privacy

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Related to https://github.com/woocommerce/woocommerce/issues/52037

This PR updates the remote logging feature to address privacy concerns. The changes are:

1. Only log errors caused by WooCommerce, rather than logging all errors containing a WooCommerce file in the trace.
2. Redact potential user data such as email addresses, phone numbers, credit card number and IP.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Create a fresh site using this branch
2. Go to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` > `Remote Logging` and enable remote logging
3. Modify [`load_rest_api`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-woocommerce.php#L538-L541) to throw an exception with sensitive data

```php
throw new Exception( 'Test exception with real sensitive data: john@example.com 555-123-4567 192.168.1.1 1234 5678 9012 3456' );
```

4. Go to Logstash page PCYsg-11Xv-p2 and confirm that the sensitive data is redacted

`Uncaught Exception: Test exception with real sensitive data: [redacted_email] [redacted_phone] [redacted_ip] [redacted_credit_card] in ./woocommerce/includes/class-woocommerce.php....`

5. Revert the changes to `load_rest_api`
6. On remote logging page, Reset the PHP rate limit
7. Modify [`woocommerce-beta-tester.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-beta-tester/woocommerce-beta-tester.php#L132-L141) to throw an exception that is not directly caused by WooCommerce but contains WooCommerce files in the trace

```php
function test_throw_error() {
	throw new Exception( 'Test PHP exception from WooCommerce Beta Tester' );
}

add_action(
	'before_woocommerce_init',
	function() {
		if ( ! get_option( 'wc_simulated_error' ) ) {
			add_option( 'wc_simulated_error', 'yes' );
			test_throw_error();
		}

		if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'product_block_editor', __FILE__, true );
		}
	}
);
```

8. Confirm that the error is not logged



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
